src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.MethodParametersInfo.cs

```diff
@@ -0,0 +1,433 @@
+            /// }
+            /// void Callee(int a, int b)
+            /// {
+            ///     DoSomething(a, b);
+            /// }
+            /// After:
+            /// void Caller(bool x)
+            /// {
+            ///     int a = Foo();
+            ///     int b = x ? Foo() : Bar();
+            ///     DoSomething(a, b);
+            /// }
+            /// void Callee(int a, int b)
+            /// {
+            ///     DoSomething(a, b);
+            /// }
+            /// </summary>
+            public ImmutableArray<(IParameterSymbol parameterSymbol, TExpressionSyntax expression)> ParametersToGenerateFreshVariablesFor { get; }
+
+            /// <summary>
+            /// A dictionary that contains Parameter that should be directly replaced. Key is the parameter and Value is the replacement exprssion
+            /// It includes
+            /// 1. Parameter mapping to literal expression
+            /// 2. Parameter that has default value, and it has no argument. It should be replaced by the default value.
+            /// 3. Parameter mapping to identifier expression
+            /// Before:
+            /// void Caller(int i, int j, bool[] k)
+            /// {
+            ///     Callee(i, j, k);
+            /// }
+            /// void Callee(int a, int b, params bool[] c)
+            /// {
+            ///     DoSomething(a, b, c);
+            /// }
+            /// After:
+            /// void Caller(int i, int j, bool[] k)
+            /// {
+            ///     DoSomething(i, j, k);
+            /// }
+            /// void Callee(int a, int b, params bool[] c)
+            /// {
+            ///     DoSomething(a, b, c);
+            /// }
+            /// 4. A special case, the parameter is only read once in the callee method body
+            /// Before:
+            /// void Caller(bool x)
+            /// {
+            ///     Callee(Foo(), Bar())
+            /// }
+            /// void Callee(int a, int b)
+            /// {
+            ///     DoSomething(a, b);
+            /// }
+            /// After:
+            /// void Caller(bool x)
+            /// {
+            ///     DoSomething(Foo(), Bar());
+            /// }
+            /// void Callee(int a, int b)
+            /// {
+            ///     DoSomething(a, b);
+            /// }
+            /// In this case, parameters 'a' and 'b' should just be replaced by the argument expression.
+            /// Note: this might cause semantics changes. It is by design.
+            /// </summary>
+            public ImmutableDictionary<IParameterSymbol, TExpressionSyntax> ParametersToReplace { get; }
+
+            public MethodParametersInfo(
+                ImmutableArray<(IParameterSymbol parameterSymbol, string name)> parametersWithVariableDeclarationArgument,
+                ImmutableArray<(IParameterSymbol parameterSymbol, TExpressionSyntax expression)> parametersToGenerateFreshVariablesFor,
```

Comment:
nit.  consider `TexpressionSyntax initExpression` to make it clearer what the expression means in the context of a "Fresh varaible".